### Extracting reads by read length
More detailed analysis of reads of specific lengths required extracting the reads of read length ranges. First, we parsed FASTQ to FASTA.
```ShellSession
# generating FASTA read file
seqkit fq2fa RSB15248-1.trimmed.fastq -o RSB15248-1.trimmed.fa
# simplifying FASTA headers
sed 's/[[:space:]].*//g' RSB15248-1.trimmed.fa | sed 's/\:/\_/g' > RSB15248-1.trimmed.fasta
```
Then, used the following `Python` script for extracting reads of specific length integers. 
```Python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed Jul  7 10:20:18 2021

@author: mansisingh
"""

#import packages 
from Bio import SeqIO

#to make the text file
text_file1 = open("RSB15248-1_2122.fa","w")
text_file2 = open("RSB15248-1_3132.fa","w")

#select sequences based on their length 
fasta_sequences = SeqIO.parse(open('RSB15248-1.trimmed.fasta'),'fasta')
for fasta in fasta_sequences:
        name, sequence = fasta.id, str(fasta.seq)
        if len(sequence) == 21 or len(sequence) == 22:
            text_file1.write(">" + name + "\n" + sequence + "\n")
            
        if len(sequence) == 31 or len(sequence) == 32:
          	text_file2.write(">" + name + "\n" + sequence + "\n")

text_file1.close()
text_file2.close()
```
The process was repeated for all samples, RSB15248-1 to RSB15248-18. Finally, we used `cat` to combine the replicate files.
```ShellSession
# reads 21-22 nt length
cat RSB15248-1_2122.fasta RSB15248-2_2122.fasta RSB15248-3_2122.fasta > Epi2122.fasta
cat RSB15248-4_2122.fasta RSB15248-5_2122.fasta RSB15248-6_2122.fasta > Hau2122.fasta
cat RSB15248-7_2122.fasta RSB15248-8_2122.fasta RSB15248-9_2122.fasta > P402122.fasta
cat RSB15248-10_2122.fasta RSB15248-11_2122.fasta RSB15248-15_2122.fasta > EVplus2122.fasta
cat RSB15248-12_2122.fasta RSB15248-13_2122.fasta RSB15248-14_2122.fasta > EVminus2122.fasta
cat RSB15248-16_2122.fasta RSB15248-17_2122.fasta RSB15248-18_2122.fasta > Myc2122.fasta
# reads 31-32 nt length
cat RSB15248-1_3132.fasta RSB15248-2_3132.fasta RSB15248-3_3132.fasta > Epi3132.fasta
cat RSB15248-4_3132.fasta RSB15248-5_3132.fasta RSB15248-6_3132.fasta > Hau3132.fasta
cat RSB15248-7_3132.fasta RSB15248-8_3132.fasta RSB15248-9_3132.fasta > P403132.fasta
cat RSB15248-10_3132.fasta RSB15248-11_3132.fasta RSB15248-15_3132.fasta > EVplus3132.fasta
cat RSB15248-12_3132.fasta RSB15248-13_3132.fasta RSB15248-14_3132.fasta > EVminus3132.fasta
cat RSB15248-16_3132.fasta RSB15248-17_3132.fasta RSB15248-18_3132.fasta > Myc3132.fasta
```

### Deduplicating reads
For better handling of the file size by collapsing identical reads into a single entry with a count value, we used `clumpify.sh` from the `BBmap` utility. We ran this line for all FASTA files above. 
```ShellSession
clumpify.sh in=Myc3132.fasta out=Myc3132_dedup.fa dedupe addcount
```
The output looks like this, containing the copy number of reads in the `copies=` suffix of the read name. 
```ShellSession
$ grep '=' Myc3132_dedup.fa | head 
>NB551536_74_H7CVKBGXF_4_12401_6241_1695_copies=2
>NB551536_74_H7CVKBGXF_4_13402_17980_20149_copies=2
>NB551536_74_H7CVKBGXF_1_23108_14983_10310_copies=10
>NB551536_74_H7CVKBGXF_1_11203_23613_16582_copies=18
>NB551536_74_H7CVKBGXF_1_21106_13537_13755_copies=3
>NB551536_74_H7CVKBGXF_1_12105_17811_18613_copies=5
```

### Obtaining RFAM databases 
We downloaded the FASTA files containing conserved ribosomal RNA sequences from [https://rfam.xfam.org](http://ftp.ebi.ac.uk/pub/databases/Rfam/CURRENT/fasta_files/)
```ShellSession
wget http://ftp.ebi.ac.uk/pub/databases/Rfam/CURRENT/fasta_files/RF00001.fa.gz
wget http://ftp.ebi.ac.uk/pub/databases/Rfam/CURRENT/fasta_files/RF00002.fa.gz
wget http://ftp.ebi.ac.uk/pub/databases/Rfam/CURRENT/fasta_files/RF00005.fa.gz
wget http://ftp.ebi.ac.uk/pub/databases/Rfam/CURRENT/fasta_files/RF01960.fa.gz
wget http://ftp.ebi.ac.uk/pub/databases/Rfam/CURRENT/fasta_files/RF02543.fa.gz
```

### Creating `MMSeqs2` databases
To run `MMSeqs2`, we first created the required `MMSeqs2` databases, which the tool requires for sequence alignment search. 
```ShellSession
# RFAM MMSeqs2 databases
mmseqs createdb RF00001.fa rf00001db
mmseqs createindex rf00001db --search-type 3
mmseqs createdb RF00002.fa rf00002db
mmseqs createindex rf00002db --search-type 3
mmseqs createdb RF00005.fa rf00005db
mmseqs createindex rf00005db --search-type 3
mmseqs createdb RF01960.fa rf01960db
mmseqs createindex rf01960db --search-type 3
mmseqs createdb RF02543.fa rf02543db
mmseqs createindex rf02543db --search-type 3
# EV minus 31-32 nt
mmseqs createdb EVminus3132_dedup.fa EVminusDB 
mmseqs createindex EVminusDB tmp --search-type 3
# EV plus 31-32 nt
mmseqs createdb EVplus3132_dedup.fa EVplusDB 
mmseqs createindex EVplusDB tmp --search-type 3
# Epi 31-32 nt
mmseqs createdb Epi3132_dedup.fa EpiDB 
mmseqs createindex EpiDB tmp --search-type 3
# Myc 31-32 nt
mmseqs createdb Myc3132_dedup.fa MycDB 
mmseqs createindex MycDB tmp --search-type 3
# P40 31-32 nt
mmseqs createdb P403132_dedup.fa P40DB 
mmseqs createindex P40DB tmp --search-type 3
# Hau 31-32 nt
mmseqs createdb Hau3132_dedup.fa HauDB 
mmseqs createindex HauDB tmp --search-type 3
```

### Running `MMSeqs2` against RFAM databases
With the databases set up, we ran `MMSeqs2`. The following scripts were run for all sample and RFAM databases above. 
```ShellSession
# run mmseqs
mmseqs search --threads 48 --search-type 3 -s 4.0 MycDB rf00001db Myc_rf00001DB tmp
# create readable result table
mmseqs convertalis MycDB rf00001db MycRfamDB Myc_rf00001DB.m8
# add an extra column to indicate which rfam database was searched, e.g.
sed -i 's/$/\trf00001/' Myc_rf00001DB.m8
# remove large search database to free up space
rm Myc_rf00001DB*
```
Because reads can have several hits in the database, we reduced the table to retain only the best hit for each read. This reduced the file size and enabled downstream analysis. The script first extracts reads that align with *Blumeria graminis* f.sp. *hordei* (CAUH01 genome accession in RFAM) or *Hordeum vulgare* (CAJW01 genome accession in RFAM). 
```ShellSession
# first for ribosomal RNAs in RF00001, RF00002, RF01960, and RF02543 in B. graminis f.sp. hordei or H. vulgare
cat <(grep CAUH01 Myc_rf00001.m8) <(grep CAJW Myc_rf00001.m8) <(grep CAUH01 Myc_rf00002.m8) <(grep CAJW Myc_rf00002.m8) <(grep CAUH01 Myc_rf01960.m8) <(grep CAJW Myc_rf01960.m8) <(grep CAUH01 Myc_rf02543.m8) <(grep CAJW Myc_rf02543.m8) | sort -k1,1 -k11r,11 > Myc_rRNA_hvubgh.m8;
awk '!visited[$1]++' Myc_rRNA_hvubgh.m8 > Myc_rRNA_hvubgh.besthit.m8
# extracting additional reads aligning with ribosomal RNA from other organisms
awk '{print $1}' Myc_rRNA_hvubgh.besthit.m8 | grep -Fvf - <(cat Myc_rf00001.m8 Myc_rf00002.m8 Myc_rf01960.m8 Myc_rf02543.m8) | sort -k1,1 -k11r,11 | awk '!visited[$1]++' > Myc_rRNA_other.besthit.m8
# then extracting reads aligning with tRNA from B. graminis f.sp. hordei or H. vulgare
cat <(grep CAUH01 Myc_rf00005.m8) <(grep CAJW Myc_rf00005.m8) | sort -k1,1 -k11r,11 > Myc_tRNA_hvubgh.m8
awk '!visited[$1]++' Myc_tRNA_hvubgh.m8 > Myc_tRNA_hvubgh.besthit.m8
```

### Extracting quantitative information from besthit tables
We used a combination of scripts to extract read counts aligning with the respective RFAM family, separately for reads from *Hordeum vulgare*, *Blumeria graminis* f.sp. *hordei*, and others. 
```ShellSession
# The number of reads aligning to the respective rRNA/tRNA is the sum of the two lines provided in each case. 
# Number of reads aligning with 5S rRNA of H. vulgare
grep CAJW Myc_rRNA_hvubgh.besthit.m8 | grep rf00001 | sed 's/\=/\t/g' | awk '{SUM+=$2}END{print SUM}'
grep CAJW Myc_rRNA_hvubgh.besthit.m8 | grep rf00001 | grep -v copies | awk '!x[$0]++' | awk 'BEGIN{i=0}{i++;}END{print i}'
# Number of reads aligning with 5.8S rRNA of H. vulgare
grep CAJW Myc_rRNA_hvubgh.besthit.m8 | grep rf00002 | sed 's/\=/\t/g' | awk '{SUM+=$2}END{print SUM}'
grep CAJW Myc_rRNA_hvubgh.besthit.m8 | grep rf00002 | grep -v copies | awk '!x[$0]++' | awk 'BEGIN{i=0}{i++;}END{print i}'
# Number of reads aligning with 18S rRNA of H. vulgare
grep CAJW Myc_rRNA_hvubgh.besthit.m8 | grep rf01960 | sed 's/\=/\t/g' | awk '{SUM+=$2}END{print SUM}'
grep CAJW Myc_rRNA_hvubgh.besthit.m8 | grep rf01960 | grep -v copies | awk '!x[$0]++' | awk 'BEGIN{i=0}{i++;}END{print i}'
# Number of reads aligning with 28S rRNA of H. vulgare
grep CAJW Myc_rRNA_hvubgh.besthit.m8 | grep rf02543 | sed 's/\=/\t/g' | awk '{SUM+=$2}END{print SUM}'
grep CAJW Myc_rRNA_hvubgh.besthit.m8 | grep rf02543 | grep -v copies | awk '!x[$0]++' | awk 'BEGIN{i=0}{i++;}END{print i}'
#
# Number of reads aligning with 5S rRNA of B. graminis f.sp. hordei
grep CAUH Myc_rRNA_hvubgh.besthit.m8 | grep rf00001 | sed 's/\=/\t/g' | awk '{SUM+=$2}END{print SUM}'
grep CAUH Myc_rRNA_hvubgh.besthit.m8 | grep rf00001 | grep -v copies | awk '!x[$0]++' | awk 'BEGIN{i=0}{i++;}END{print i}'
# Number of reads aligning with 5.8S rRNA of B. graminis f.sp. hordei
grep CAUH Myc_rRNA_hvubgh.besthit.m8 | grep rf00002 | sed 's/\=/\t/g' | awk '{SUM+=$2}END{print SUM}'
grep CAUH Myc_rRNA_hvubgh.besthit.m8 | grep rf00002 | grep -v copies | awk '!x[$0]++' | awk 'BEGIN{i=0}{i++;}END{print i}'
# Number of reads aligning with 18S rRNA of B. graminis f.sp. hordei
grep CAUH Myc_rRNA_hvubgh.besthit.m8 | grep rf01960 | sed 's/\=/\t/g' | awk '{SUM+=$2}END{print SUM}'
grep CAUH Myc_rRNA_hvubgh.besthit.m8 | grep rf01960 | grep -v copies | awk '!x[$0]++' | awk 'BEGIN{i=0}{i++;}END{print i}'
# Number of reads aligning with 28S rRNA of B. graminis f.sp. hordei
grep CAUH Myc_rRNA_hvubgh.besthit.m8 | grep rf02543 | sed 's/\=/\t/g' | awk '{SUM+=$2}END{print SUM}'
grep CAUH Myc_rRNA_hvubgh.besthit.m8 | grep rf02543 | grep -v copies | awk '!x[$0]++' | awk 'BEGIN{i=0}{i++;}END{print i}'
# Number of reads aligning with tRNAs of B. graminis f.sp. hordei
grep CAUH Myc_rRNA_hvubgh.besthit.m8 | grep rf00005 | sed 's/\=/\t/g' | awk '{SUM+=$2}END{print SUM}'
grep CAUH Myc_rRNA_hvubgh.besthit.m8 | grep rf00005 | grep -v copies | awk '!x[$0]++' | awk 'BEGIN{i=0}{i++;}END{print i}'
#
# Number of reads aligning rRNA of neither H. vulgare nor B. graminis f.sp. hordei
# 5S rRNA
grep rf00001 Myc_rRNA_other.besthit.m8 | sed 's/\=/\t/g' | awk '{SUM+=$2}END{print SUM}' 
grep rf00001 Myc_rRNA_other.besthit.m8 | grep -v copies | awk '!x[$0]++' | awk 'BEGIN{i=0}{i++;}END{print i}'
# 5.8S rRNA
grep rf00002 Myc_rRNA_other.besthit.m8 | sed 's/\=/\t/g' | awk '{SUM+=$2}END{print SUM}'
grep rf00002 Myc_rRNA_other.besthit.m8 | grep -v copies | awk '!x[$0]++' | awk 'BEGIN{i=0}{i++;}END{print i}'
# 18S rRNA
grep rf01960 Myc_rRNA_other.besthit.m8 | sed 's/\=/\t/g' | awk '{SUM+=$2}END{print SUM}'
grep rf01960 Myc_rRNA_other.besthit.m8 | grep -v copies | awk '!x[$0]++' | awk 'BEGIN{i=0}{i++;}END{print i}'
# 28S rRNA
grep rf02543 Myc_rRNA_other.besthit.m8 | sed 's/\=/\t/g' | awk '{SUM+=$2}END{print SUM}'
grep rf02543 Myc_rRNA_other.besthit.m8 | grep -v copies | awk '!x[$0]++' | awk 'BEGIN{i=0}{i++;}END{print i}'
```

### Read alignment distribution along the 28S rRNAs of *Hordeum vulgare* and *Blumeria graminis* f.sp. *hordei*
We isolated the reads mapping to the 28S rRNA (RF02543) to further analyze. 
```ShellSession
grep rf02543 Myc_rRNA_hvubgh.besthit.m8 | awk '{print $1}' | seqkit grep -n -f - Myc3132_dedup.fa > Myc3132_dedup.rf02543.fasta
```
Then, we needed new `MMSeqs2` databases, here shown for Myc samples. The same databases were created for the other samples accordingly. In addition, we created databases for the *Hordeum vulgare* 28S rRNA [URS0002132C2A_4513](https://rnacentral.org/rna/URS0002132C2A/4513) and the *B. graminis* f.sp. *hordei* 28S rRNA [URS0002174482_62688](https://rnacentral.org/rna/URS0002174482/62688). 
```ShellSession
# create reads DB
mmseqs createdb Myc3132_dedup.rf02543.fasta MycDB_rf02543
mmseqs createindex MycDB_rf02543 tmp --search-type 3
# create 28S rRNA DBs
mmseqs createdb URS0002174482_62688.fasta Bgh28SDB
mmseqs createindex Bgh28SDB tmp --search-type 3
mmseqs createdb URS0002132C2A_4513.fasta Hvu28SDB
mmseqs createindex Hvu28SDB tmp --search-type 3
```
Then, we reran `MMSeqs2` of the 28S rRNA-derived reads against the 28S rRNA sequences. 
```ShellSession
# run mmseqs
mmseqs search --threads 48 --search-type 3 -s 4.0 MycDB_rf02543 Bgh28SDB MycDB_28S_Bgh tmp
mmseqs search --threads 48 --search-type 3 -s 4.0 MycDB_rf02543 Hvu28SDB MycDB_28S_Hvu tmp
# create readable result table
mmseqs convertalis MycDB_rf02543 Bgh28SDB MycDB_28S_Bgh MycDB_28S_Bgh.m8
mmseqs convertalis MycDB_rf02543 Hvu28SDB MycDB_28S_Hvu MycDB_28S_Hvu.m8
# remove large search databases to free up space
rm MycDB_28S_Bgh* MycDB_28S_Hvu*
# filter
sort -k1,1 -k11r,11 MycDB_28S_Bgh.m8 | awk '!visited[$1]++' > MycDB_28S_Bgh.besthit.m8
sort -k1,1 -k11r,11 MycDB_28S_Hvu.m8 | awk '!visited[$1]++' > MycDB_28S_Hvu.besthit.m8
```
Finally, we created count/position tables towards plotting the data. We generated 10-bp non-overlapping sliding window BED files of the 28S rRNA sequences first.
```
$ cat URS0002174482_62688.txt
Bgh_28S 3564
$ cat URS0002132C2A_4513.txt
Hvulgare_28S    3853
```
```ShellSession
bedtools makewindows -g URS0002174482_62688.txt -w 10 > URS0002174482_62688.bed
bedtools makewindows -g URS0002132C2A_4513.txt -w 10 > URS0002132C2A_4513.bed
```
The resulting BED files look like this:
```
$ head URS0002174482_62688.bed
Bgh_28S	0	10
Bgh_28S	10	20
Bgh_28S	20	30
Bgh_28S	30	40
Bgh_28S	40	50
Bgh_28S	50	60
Bgh_28S	60	70
Bgh_28S	70	80
Bgh_28S	80	90
Bgh_28S	90	100
```
Then, we were able to create read count coverage across the 28S rRNA sequences in 10-bp windows via `BEDtools`. 
```ShellSession
# Parsing text files into BED files
awk -F"\t" '{ print $2,$9,$10,$1 }' MycDB_28S_Bgh.besthit.m8 | sed 's/$/\t1/' | sed 's/_copies\=/\t/g' | sort -k2,2n | awk '{ print $1,$2,$3,$4,$5 }' | sed 's/[[:space:]]/\t/g' > MycDB_28S_Bgh.bed
awk -F"\t" '{ print $2,$9,$10,$1 }' MycDB_28S_Hvu.besthit.m8 | sed 's/$/\t1/' | sed 's/_copies\=/\t/g' | sort -k2,2n | awk '{ print $1,$2,$3,$4,$5 }' | sed 's/[[:space:]]/\t/g' > MycDB_28S_Hvu.bed
# Using BEDtools to create read count coverage for 10-bp windows
bedtools map -c 5 -o sum -a URS0002174482_62688.bed -b MycDB_28S_Bgh.bed > MycDB_28S_Bgh.count.bed
bedtools map -c 5 -o sum -a URS0002132C2A_4513.bed -b MycDB_28S_Hvu.bed > MycDB_28S_Hvu.count.bed
# since bedtools places a dot for zeros, we replaced the dots
sed -i 's/\t\./\t0/g' *.count.bed
```
The resulting tables will look as shown below, where column 5 contains the counts of alignments for each 10-bp window. 
```
$ head MycDB_28S_Hvu.count.bed
Hvulgare_28S	1	10	a1	0
Hvulgare_28S	11	20	a2	0
Hvulgare_28S	21	30	a3	0
Hvulgare_28S	31	40	a4	0
Hvulgare_28S	41	50	a5	0
Hvulgare_28S	51	60	a6	0
Hvulgare_28S	61	70	a7	0
Hvulgare_28S	71	80	a8	0
Hvulgare_28S	81	90	a9	0
Hvulgare_28S	91	100	a10	0
```
